### ISSUE TITLE
Stackoverflow error with decision tree

### ISSUE THREAD

**Issue by tlienart** [alan-turing-institute/MLJModels.jl#126]
```julia
using MLJ
X, y = @load_iris
@load DecisionTreeClassifier
tree = machine(DecisionTreeClassifier(), X, y)
fit!(tree)
```

Stack overflows on 1.3 :

```
[ Info: Training Machine{DecisionTreeClassifier} @ 9…30.
ERROR: StackOverflowError:
Stacktrace:
 [1] ht_keyindex(::Dict{Any,UInt8}, ::SubString{String}) at ./dict.jl:289
 [2] in(::SubString{String}, ::Base.KeySet{Any,Dict{Any,UInt8}}) at ./dict.jl:546
 [3] issubset(::Array{Any,1}, ::Base.KeySet{Any,Dict{Any,UInt8}}) at ./abstractset.jl:277
 [4] ⊈ at ./abstractset.jl:341 [inlined]
 [5] get!(::CategoricalArrays.CategoricalPool{Any,UInt8,CategoricalArrays.CategoricalValue{Any,UInt8}}, ::CategoricalArrays.CategoricalValue{Any,UInt8}) at /Users/tlienart/.julia/packages/CategoricalArrays/kxnZh/src/pool.jl:197
 [6] get!(::CategoricalArrays.CategoricalPool{Any,UInt8,CategoricalArrays.CategoricalValue{Any,UInt8}}, ::CategoricalArrays.CategoricalValue{Any,UInt8}) at /Users/tlienart/.julia/packages/CategoricalArrays/kxnZh/src/pool.jl:204 (repeats 25976 times)
 [7] setindex!(::CategoricalArrays.CategoricalArray{Any,1,UInt8,Any,CategoricalArrays.CategoricalValue{Any,UInt8},Missing}, ::CategoricalArrays.CategoricalValue{Any,UInt8}, ::Int64) at /Users/tlienart/.julia/packages/CategoricalArrays/kxnZh/src/array.jl:337
 [8] collect_to_with_first! at ./array.jl:659 [inlined]
 [9] collect(::Base.Generator{Array{Any,1},MLJBase.var"#19#20"{CategoricalArrays.CategoricalPool{Any,UInt8,CategoricalArrays.CategoricalValue{Any,UInt8}}}}) at ./array.jl:641
 [10] fit(::DecisionTreeClassifier, ::Int64, ::NamedTuple{(:sepal_length, :sepal_width, :petal_length, :petal_width),NTuple{4,Array{Float64,1}}}, ::CategoricalArrays.CategoricalArray{Any,1,UInt8,Any,CategoricalArrays.CategoricalValue{Any,UInt8},Missing}) at /Users/tlienart/.julia/dev/MLJBase/src/data.jl:141
 [11] fit!(::Machine{DecisionTreeClassifier}; rows::Nothing, verbosity::Int64, force::Bool) at /Users/tlienart/.julia/dev/MLJ/src/machines.jl:142
 [12] fit!(::Machine{DecisionTreeClassifier}) at /Users/tlienart/.julia/dev/MLJ/src/machines.jl:111
 [13] top-level scope at REPL[5]:1
```

On 1.2 it's terser but even more problematic

```
julia> fit!(tree)
[ Info: Training Machine{DecisionTreeClassifier} @ 1…89.
Bus error: 10
[~] > 
```

**edit** long story short it’s caused by the new way basic data is loaded where more of the elements are default typed to Any; this causes issues when multiclass; there are two ways forward (not mutually exclusive)

1. Fully type each column before coercion in the datasets, this is easy and I’ll try to provide a patch soon on MLJBase
2. Stuff should not fail like it does now; one way to prevent this is ensure that the coerce to multiclass of something that’s currently machine typed as any should be re-typed as string; I’ve got a patch for this; I’ll open a PR for this on ST

As an aside I don’t think we should revert the delimitedfiles change; it makes more sense to have a very bare bone data thing by default and making sure that we can deal with it is important given we claim to support all tabular data and not just dataframes. 

Another note is that this suggests having a procedure for integrated testing; but in fact the tutorials may just be that; currently they fail bc of the issue discussed here but otherwise they provide a nice way to check that things work together as expected

**Comment by tlienart** [comment]
So  I downgraded to 0.5.5 but it still fails, am trying to see if it's something to do with the latest release of MLJ; I think this kind of issue supports https://github.com/alan-turing-institute/MLJ.jl/issues/323

**Comment by tlienart** [comment]
Ok so no, the Manifest file [here](https://github.com/alan-turing-institute/MLJTutorials/blob/1e5240f2214b14e458a4b8926304ba45e0d4db59/Manifest.toml) suggests that MLJ 0.5.2 + MLJModels 0.5.5 + MLJBase 0.7.3 works; trying to revert that

**Comment by tlienart** [comment]
Right so reverting MLJBase to  0.7.3 fixes the issue, this points strongly towards  https://github.com/alan-turing-institute/MLJBase.jl/pull/79 being the culprit; @ablaom could you look into this asap?

**Comment by tlienart** [comment]
Hmm this seems to be a type issue after the `coerce` to ScientificTypes; the problem is now that `y` is

**now**
```
CategoricalArrays.CategoricalArray{Any,1,UInt8}
```

**before**
```
CategoricalArrays.CategoricalArray{String,1,UInt32}
```

However that shouldn't cause a stack overflow; `levels` correctly leads to the right levels. 

===

Ok this has to do with how/when things are parsed:

```julia-repl
julia> typeof(y.pool.levels)
Array{Any,1}
```

but

```julia-repl
julia> [v for v in y.pool.levels]
3-element Array{SubString{String},1}:
 "setosa"
 "versicolor"
 "virginica"
```

and then the issue has to do with the `valindex / invindex` thing. I'll try to find a fix.

**Comment by ablaom** [comment]
Actually a bug in CategoricalArrays but is resolved by a workaround implemented in ScientificTypes 0.7.2, release pending : https://github.com/JuliaRegistries/General/pull/5267
